File: examples/c/src/shape.c
#include "shape.h"

const char *color_name(Color c) {
    switch (c) {
        case COLOR_RED:    return "Red";
        case COLOR_GREEN:  return "Green";
        case COLOR_BLUE:   return "Blue";
        case COLOR_YELLOW: return "Yellow";
        case COLOR_WHITE:  return "White";
    }
    return "Unknown";
}

double shape_area(const Shape *s) {
    return s->vtable->area(s);
}

double shape_perimeter(const Shape *s) {
    return s->vtable->perimeter(s);
}

Vec3 shape_centroid(const Shape *s) {
    return s->vtable->centroid(s);
}

void shape_describe(const Shape *s, FILE *out) {
    Vec3 c = shape_centroid(s);
    fprintf(out, "[%s] %s\n", s->name, color_name(s->color));
    fprintf(out, "  area:      %.2f\n", shape_area(s));
    fprintf(out, "  perimeter: %.2f\n", shape_perimeter(s));
    fprintf(out, "  centroid:  (%.2f, %.2f, %.2f)\n", c.x, c.y, c.z);
}

void shape_destroy(Shape *s) {
    if (s && s->vtable->destroy) {
        s->vtable->destroy(s);
    }
}

File: examples/rust/src/shape.rs
use crate::vector3::Vec3;
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Color {
    Red,
    Green,
    Blue,
    Yellow,
    White,
}

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Color::Red    => write!(f, "Red"),
            Color::Green  => write!(f, "Green"),
            Color::Blue   => write!(f, "Blue"),
            Color::Yellow => write!(f, "Yellow"),
            Color::White  => write!(f, "White"),
        }
    }
}

pub trait Shape {
    fn name(&self) -> &str;
    fn color(&self) -> Color;
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;
    fn centroid(&self) -> Vec3;

    fn describe(&self) -> String {
        let c = self.centroid();
        format!(
            "[{}] {}\n  area:      {:.2}\n  perimeter: {:.2}\n  centroid:  ({:.2}, {:.2}, {:.2})",
            self.name(), self.color(),
            self.area(), self.perimeter(),
            c.x, c.y, c.z,
        )
    }
}
